### PR TITLE
Fix AsuraScans

### LINKED ISSUE #4167: Asura Scans site rewritten
### Source information

Asura Scans 1.4.34

### Source language

English

### Steps to reproduce

**Scenario 1**
1. Navigate to the "Sources" page in the "Browse" tab.
2. Select "Asura Scans"
3. Complete the CloudFlare captcha in WebView
4. Exit WebView
5. Click the "Refresh" button

OR

**Scenario 2**
1. Open any manhwa saved in the library from "Asura Scans"
2. Complete the CloudFlare captcha in WebView
3. Exit WebView
4. Attempt to refresh manhwa for new chapters

### Expected behavior

**Scenario 1**
List of all manhwa to be shown with titles and cover images

**Scenario 2**
Number of chapters to update if there are new chapters else nothing changes

### Actual behavior

**Scenario 1**
A blank screen with "No results found" is shown

**Scenario 2**
Displays the loading spinner for a long time then shows a toast notification saying "Unknown error"

### Mihon/Tachiyomi version

Mihon v1.16.5 | TachiJ2K 1.7.4

### Android version

Android 13

### Other details

The Asura Scans site seems to have gotten a complete rewrite using Next.js and Tailwindcss.

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a short but informative title.
- [X] I have updated the app to version **[0.15.3](https://github.com/mihonapp/mihon/releases/latest)**.
- [X] I have updated all installed extensions.
- [X] I have tried the [troubleshooting guide](https://tachiyomi.org/help/guides/troubleshooting/).
- [x] If this is an issue with the app itself, I should be opening an issue in the [app repository](https://github.com/tachiyomiorg/tachiyomi/issues/new/choose).
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by bapeey**
Dynamic URL relies on the redirect that asura does, can be changed to a search based in the future

Closes #4167 

Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [ ] Have removed `web_hi_res_512.png` when adding a new extension

### CODE REVIEW

**Comment by lamaxama** [comment]
Thank you for your work. I've tested it, and it works. Since dynamic URLs have been added, can we further add an interceptor to handle the 404 case and automatically migrate, like [MangaThemesiaAlt](https://github.com/keiyoushi/extensions-source/blob/682afcabce4353c281e53fc2319ee5c826613d05/lib-multisrc/mangathemesia/src/eu/kanade/tachiyomi/multisrc/mangathemesia/MangaThemesiaAlt.kt#L152)? In that scenario, most users won't need to migrate a ton of titles in the library.

If we add search-based dynamic urls in the future, then existing titles can reuse this ability to automatically migrate. 
However, if we change the versionId, then that opportunity will be lost forever. 
Please don't modify the versionId.

**Review by hemanthmannam — COMMENTED** [review]
All good

**Review by Lygaen — COMMENTED** [review]
Works for me as well!

**Comment by Lygaen** [comment]
It seems that adding might just work, I'll try it on my machine

```kt

        .addInterceptor { chain ->
            val request = chain.request()
            val response = chain.proceed(request)

            if (!response.isSuccessful && response.code == 404) {
                response.close()

                val dynUrl = request.url.toString().toDynamicUrlIfNeeded()

                val newRequest = request.newBuilder()
                    .url(dynUrl)
                    .build()

                return@addInterceptor chain.proceed(newRequest)
            }

            return@addInterceptor response
        }

```

**Comment by hemanthmannam** [comment]
Someone please approve this

**Comment by Lygaen** [comment]
> It seems that adding might just work, I'll try it on my machine
> 
> ```kotlin
>         .addInterceptor { chain ->
>             val request = chain.request()
>             val response = chain.proceed(request)
> 
>             if (!response.isSuccessful && response.code == 404) {
>                 response.close()
> 
>                 val dynUrl = request.url.toString().toDynamicUrlIfNeeded()
> 
>                 val newRequest = request.newBuilder()
>                     .url(dynUrl)
>                     .build()
> 
>                 return@addInterceptor chain.proceed(newRequest)
>             }
> 
>             return@addInterceptor response
>         }
> ```

It works ! Manual intervention for migration will be required nonetheless but at least the interceptor works.

The migration will need to be done manually because the mangas reference a deleted extension on update, so, even with an interceptor, users will need to migrate manually

**Comment by Smol-Ame** [comment]
> Someone please approve this

Don't rush, all the work is done by volunteers using their own time. Wait for it to be reviewed/approved, or migrate to a different source.

**Comment by liven-maksim-infotecs** [comment]
> users will need to migrate manually

What do you mean by that? 
1. by using migration button 
2. add new version of manga manually and than remove old one

**Comment by Lygaen** [comment]
No they will need to push the migrate button

File: src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt [hunk 1]
```diff
@@ -1,21 +1,52 @@
...
+        Injekt.get<Application>().getSharedPreferences("source_$id", 0x0000)
 
-class AsuraScans : MangaThemesiaAlt(
-    "Asura Scans",
-    "https://asuracomic.net",
-    "en",
-    dateFormat = SimpleDateFormat("MMM d, yyyy", Locale.US),
-    randomUrlPrefKey = "pref_permanent_manga_url_2_en",
-) {
     init {
```
Comment (AwkwardPeak7) [inline src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt:48]: This won't work as the id will have changed, you can directly delete the old shared pref by hard coding the old id.

**Review by AwkwardPeak7 — COMMENTED** [review]
I am still iffy on the URL situation. Maybe save the slug only in the dB and handle rest of the URL through the pref map. Also would it be possible to get all series slug from the API or something? (Kinda like old HeanCMS impl). We can further discuss in discord.

**Comment by MaxLevs** [comment]
> I am still iffy on the URL situation. Maybe save the slug only in the dB and handle rest of the URL through the pref map. 

But nothing will change if they not only change the base url. If changes in url path occur, then code modifications need to be done.

**Comment by lamaxama** [comment]
@bapeey  Hey, I added the automatic migration feature based on your modifications. It has been tested, and most titles can be migrated automatically.  https://github.com/lamaxama/extensions-source/commit/0c9e4806c8f7eb5e82be6eb802f0531771a402d6  and https://github.com/lamaxama/extensions-source/commit/fe0f3147314709ff1261d3704ac27ffd2123fcfa
If possible, you can simply copy the code. I just don't want to migrate tons of titles in the library.

Update: 

Additionally, the new site will automatically redirect https://asuracomic.net/series/the-lords-coins-arent-decreasing-123456 to https://asuracomic.net/series/the-lords-coins-arent-decreasing-86df2251. Therefore, having a search base dynamic URL may not be necessary.

**Comment by 1aam2am1** [comment]
Old: https://asuracomic.net/manga/1908287720-martial-god-regressed-to-level-2/
New: https://asuracomic.net/series/martial-god-regressed-to-level-2-ce6b06dd

I don't think we can do anything in this case.
We could try saving it as 'Permament manga url'/only webpage or series name as key.

But it won't work if they are changing it so much.
So I think we can try adding some way so user can change domain url. Or that plugin will try to work with single redirection.

However this should be discussion after delivering the fix. We can try future proof the plugin so in the next site change, the migration wouldn't be needed. But that's it.

**Comment by MaxLevs** [comment]
> Or that plugin will try to work with single redirection.

They are redirecting to the '/' if I recall correctly

UPD: yep. It's redirection on the index page

**Comment by bapeey** [comment]
@lamaxama i added a similar workaround for auto migration, can you test?

**Comment by lamaxama** [comment]
@bapeey It works! Truly a clever approach! One more thing, we can now remove this line: Injekt.get().getSharedPreferences("source_6247824327199706550", 0x0000), because we have not changed the versionId.

**Comment by AhYOh69** [comment]
I don't understand anything at all but it looks cool... So when and where can we download the extension 🥹?

**Comment by bellfireg** [comment]
> I don't understand anything at all but it looks cool... So when and where can we download the extension 🥹?

just wait and see . they still working on it using their time . you just need to wait.

**Comment by zephyrxk** [comment]
How to do this still showing no results found

**Comment by zephyrxk** [comment]
> > I don't understand anything at all but it looks cool... So when and where can we download the extension 🥹?
> 
> just wait and see . they still working on it using their time . you just need to wait.

Is it still not fixed the asura scan opening?

**Comment by Thaafisilt** [comment]
I never felt so dumb in my life until I saw what was happening here. It's like my brain went on vacation without telling me!

**Comment by Gamermemer** [comment]
> I never felt so dumb in my life until I saw what was happening here. It's like my brain went on vacation without telling me!

Very very true, now just waiting for this God's man to fix the repo

**Comment by Smol-Ame** [comment]
Can everybody not comment here if you aren't contributing? This pings the server channel & emails everyone in this pull request.

Wait for a fix, and if you need to discuss how badly you want the fix, do it in the Keiyoushi Discord repo-support channel instead.

File: src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt [hunk 2]
```diff
@@ -25,47 +54,271 @@ class AsuraScans : MangaThemesiaAlt(
...
+        }
+    }
+
+    override fun getMangaUrl(manga: SManga): String {
+        if (!preferences.dynamicUrl()) return super.getMangaUrl(manga)
+        val match = OLD_FORMAT_MANGA_REGEX.find(manga.url)?.groupValues?.get(1)
+        val slug = match ?: manga.url.substringAfter("/series/").substringBefore("/")
+        val savedSlug = preferences.slugMap[slug] ?: "$slug-"
+        return baseUrl + manga.url.replace(slug, savedSlug)
+    }
```
Comment (AwkwardPeak7) [inline src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt]: `mangaDetailsRequest` builds same(?) url so no need to override

File: src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt [hunk 3]
```diff
@@ -25,47 +54,271 @@ class AsuraScans : MangaThemesiaAlt(
...
+            val newSlug = url.substringAfter("/series/").substringBefore("/")
+            val absSlug = newSlug.substringBeforeLast("-")
+            preferences.slugMap = preferences.slugMap.apply { put(absSlug, newSlug) }
+        }
+        return super.mangaDetailsParse(response)
+    }
+
+    override fun mangaDetailsParse(document: Document) = SManga.create().apply {
+        title = document.selectFirst("span.text-xl.font-bold")!!.ownText()
+        thumbnail_url = document.selectFirst("img[alt=poster]")!!.attr("abs:src")
```
Comment (AwkwardPeak7) [inline src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt]: no need to add `!!` to non-essential fields

File: src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt [hunk 4]
```diff
@@ -25,47 +54,271 @@ class AsuraScans : MangaThemesiaAlt(
...
+            0L
+        }
+    }
+
+    override fun getChapterUrl(chapter: SChapter): String {
+        if (!preferences.dynamicUrl()) return super.getChapterUrl(chapter)
+        val slug = chapter.url.substringAfter("/series/").substringBefore("/")
+        val savedSlug = preferences.slugMap[slug] ?: "$slug-"
+        return baseUrl + chapter.url.replace(slug, savedSlug)
+    }
```
Comment (AwkwardPeak7) [inline src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt]: same here

File: src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt [hunk 5]
```diff
@@ -25,47 +54,271 @@ class AsuraScans : MangaThemesiaAlt(
...
+            summary = PREF_DYNAMIC_URL_SUMMARY
+            setDefaultValue(PREF_DYNAMIC_URL_DEFAULT)
+        }.let(screen::addPreference)
+    }
+
+    private var SharedPreferences.slugMap: MutableMap<String, String>
+        get() {
+            val jsonMap = getString(PREF_SLUG_MAP, "{}")!!
+            val slugMap = runCatching { json.decodeFromString<Map<String, String>>(jsonMap) }
+            return slugMap.getOrNull()?.toMutableMap() ?: mutableMapOf()
```
Comment (AwkwardPeak7) [inline src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt]:
```suggestion
            return try {
                json.decodeFromString<Map<String, String>>(jsonMap).toMutableMap()
            } catch (_: Exception) {
                mutableMapOf()
            }
```

File: src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt [hunk 6]
```diff
@@ -25,47 +54,271 @@ class AsuraScans : MangaThemesiaAlt(
...
+        if (!preferences.dynamicUrl()) return this
+        val slug = this.substringAfter("/series/").substringBefore("/")
+        val absSlug = slug.substringBeforeLast("-")
+        preferences.slugMap = preferences.slugMap.apply { put(absSlug, slug) }
+        return this.replace(slug, absSlug)
+    }
+
+    companion object {
+        private val CLEAN_DATE_REGEX = """(\d+)(st|nd|rd|th)""".toRegex()
+        private val OLD_FORMAT_MANGA_REGEX = """^/manga/\d*(.*?)/?$""".toRegex()
```
Comment (AwkwardPeak7) [inline src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt]: /manga/1223-cool-manga/
/manga/cool-manga/

```suggestion
        private val OLD_FORMAT_MANGA_REGEX = """^/manga/(\d+-)?([^/]+)/?$""".toRegex()
```

would have to adjust the match group elsewhere in the code too

File: src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt [hunk 7]
```diff
@@ -25,47 +54,271 @@ class AsuraScans : MangaThemesiaAlt(
...
+        private val OLD_FORMAT_CHAPTER_REGEX = """^/\d+-[^/]*-chapter-\d+(-\d+)*/?$""".toRegex()
```
Comment (AwkwardPeak7) [inline src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt]:
```suggestion
        private val OLD_FORMAT_CHAPTER_REGEX = """^/(\d+-)?[^/]*-chapter-\d+(-\d+)*/?$""".toRegex()
```
similar here

File: src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt [hunk 8]
```diff
@@ -25,47 +54,271 @@ class AsuraScans : MangaThemesiaAlt(
...
+        private const val PREF_SLUG_MAP = "pref_slug_map"
+        private const val PREF_DYNAMIC_URL = "pref_dynamic_url"
+        private const val PREF_DYNAMIC_URL_TITLE = "Automatically update dynamic URLs"
+        private const val PREF_DYNAMIC_URL_SUMMARY = "Automatically update random numbers in manga URLs.\nHelps mitigating HTTP 404 errors during update and \"in library\" marks when browsing.\nNote: This setting may require clearing database in advanced settings and migrating all manga to the same source."
```
Comment (AwkwardPeak7) [inline src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt]: inline these

File: src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt [hunk 9]
```diff
@@ -25,47 +54,271 @@ class AsuraScans : MangaThemesiaAlt(
...
+
+        return GET(url.build(), headers)
+    }
+
+    override fun searchMangaSelector() = "div.grid > a[href]"
+
+    override fun searchMangaFromElement(element: Element) = SManga.create().apply {
+        setUrlWithoutDomain(element.attr("abs:href").toPermSlugIfNeeded())
+        title = element.selectFirst("div.block > span.block")!!.ownText()
+        thumbnail_url = element.selectFirst("img")!!.attr("abs:src")
```
Comment (AwkwardPeak7) [inline src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt]:
```suggestion
        thumbnail_url = element.selectFirst("img")?.attr("abs:src")
```

**Review by AwkwardPeak7 — CHANGES_REQUESTED** [review]

**Comment by uChieLs** [comment]
> Works for me as well!
 
i'm sorry , i'm new and searching how to solve asura scan no result found .
can please explain how to solve it ?
i read from top to bottom but i can't understand how to .
sorry for my bad english and thanks .

**Comment by Norsze** [comment]
This is an ongoing development of the extension, if you don't have anything to contribute to the effort with programming knowledge, then please don't comment here, you won't get help here. This section is for helping the development.

**Review by AwkwardPeak7 — APPROVED** [review]

**Pull request merged**